### PR TITLE
Code quality issues identified by SonarQube

### PR DESCRIPTION
**Pull request by davidjsherman**
### Code quality
I have started running [SonarQube](https://www.sonarqube.org) on the Aseba and Enki code bases, and here is a PR to discuss the improvements to code quality that SonarQube suggests. It includes #28. Each commit in this PR addresses a separate rule; for example, 82303c75cbb9fb1282945412406176992a9d7f46 addresses rule [cpp:S3230](https://sonarcloud.io/organizations/davidjsherman-github/rules#q=cpp%3AS3230). It should be possible to cherry-pick individual commits. This PR resolves roughly half of the issues identified in the Dashel code.

### Remaining issues
The SonarQube report at https://sonarcloud.io/dashboard?id=org.thymio%3Adashel is up to date with this PR, and shows the issues that remain *after* this PR is applied. The following is a list of the most frequent rules that are still triggered. (**Note that each rule page ends with a link to specific issues**)

- [Obsolete POSIX functions should not be used](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3AObsoletePosixFunction)
- [Non-reentrant POSIX functions should be replaced with their reentrant versions](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3ANonReentrantFunction)

This rule is triggered by the complex class hierarchy in Dashel and the need for `protected` access to attributes. It might be appropriate to ignore this rule.
- [Member variables should not be "protected"](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3AS3656)

This rule can probably be ignored, since the Aseba convention is to specify access separately for attributes and methods:
- [Access specifiers should not be redundant](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3AS3539)

Finally, two functions are identified as being too complex. At a minimum unit tests (#27) need to be defined before addressing this problem.
- [Cognitive Complexity of functions should not be too high](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3AS3776)

### CODE REVIEW

**Comment by stephanemagnenat** [comment]
I am wondering why "Preprocessor directives should not be indented", if it eases the readability of code?

**Comment by stephanemagnenat** [comment]
Also, why "Init-declarator-lists and member-declarator-lists should consist of single init-declarators and member-declarators respectively"?

File: dashel/dashel-posix.cpp [hunk 1]
```diff
@@ -1090,47 +1090,6 @@ namespace Dashel
 	};
 	
 	
-	/*
```
Comment (stephanemagnenat) [inline dashel/dashel-posix.cpp]: I think this block of code should be moved into an example.

**Comment by stephanemagnenat** [comment]
Also I would use const reference for exceptions.

**Comment by stephanemagnenat** [comment]
Beside these, ok to merge.

**Comment by davidjsherman** [comment]
We can add exceptions to the rules, especially if there is already a different community standard in the code. That's why I removed the rule that namespaces should start with lower case [a-z], for example. Here is the documentation for the two rules you mentioned:
- [Init-declarator-lists and member-declarator-lists should consist of single init-declarators and member-declarators respectively](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3ASingleDeclarationPerStatement)
  - MISRA C++:2008, 8-0-1
  - [CERT, DCL04-CPP. - Do not declare more than one variable per declaration](https://www.securecoding.cert.org/confluence/x/fAAhAQ)
- [Preprocessor directives should not be indented](https://sonarcloud.io/organizations/davidjsherman-github/rules#rule_key=cpp%3APPDirectiveIndentation)

**Comment by stephanemagnenat** [comment]
For the variable declarations separated by a comma, I understand the argument that it might be confusing for some developers who are not used to this feature of C++, so I'm fine with enforcing this rule.

For _Preprocessor directives should not be indented_, I agree for the change to put the #ifdef without indentation, the rationale makes sense. However, I like the indented includes within the #ifdef, to separate them from the #ifdef itself. What is your personal opinion on that?

Also, I'm more than happy to merge the changes we agree on.

**Comment by davidjsherman** [comment]
I loathe the preprocessor and I like the rule to not indent `#ifdef`s because otherwise they give the impression that they respect C++ scoping.

Preprocessor directives with conditional logic make me nervous because I think that most of the time the configuration system is better informed. For example, [L77](https://github.com/aseba-community/dashel/blame/master/dashel/dashel-posix.cpp#L77), I should have left the decision to CMake, whether TARGET_OS_IPHONE depended on MACOSX or not. 

>However, I like the indented includes within the #ifdef, to separate them from the #ifdef itself. What is your personal opinion on that?

If I understand your question, you are referring to the conditional includes [L71](https://github.com/aseba-community/dashel/blame/master/dashel/dashel-posix.cpp#L71), [L94](https://github.com/aseba-community/dashel/blame/master/dashel/dashel-posix.cpp#L94), etc. In those cases the scoping is the preprocessor itself, and indenting doesn't give a false impression. So if it really improves readability we could manually allow exceptions to the rule in those places.

**Comment by davidjsherman** [comment]
Rereading my answer, I didn't mean to be passive-aggressive. My opinion is that if you already have a convention then there is no compelling reason to change it.

**Comment by davidjsherman** [comment]
f30e48caa35314a1c447e91e5d7bee20b0bf3da5 catches exceptions with const references

**Pull request merged**